File: android/app/src/main/java/com/llmhub/llmhub/mimobot/audio/AudioIO.kt
package com.llmhub.llmhub.mimobot.audio

import kotlinx.coroutines.flow.Flow

/**
 * Abstractions that let [com.llmhub.llmhub.mimobot.pipeline.VoicePipeline] run
 * against either the phone's own mic+speaker (local dev mode) or a BLE
 * transport with Opus codec (real device mode).
 *
 * Contract: all frames are 16 kHz mono Int16 PCM, 320 samples (20 ms) each.
 */
interface AudioSource {
    /** Start capturing; flow completes when [stop] is called. */
    fun frames(): Flow<ShortArray>
    fun stop()
}

interface AudioSink {
    /** Must be called before [write]. Idempotent. */
    fun start()
    suspend fun write(frame: ShortArray)
    fun stop()
}

File: ios/LLMHub/Sources/LLMHub/MimoBot/AudioIO.swift
import Foundation

/// Source/sink abstractions so `VoicePipeline` runs against either the phone's
/// own mic+speaker (local dev mode) or a BLE transport + Opus codec (real
/// device). All frames are 16 kHz mono Int16, `MimoBotIds.frameSamples` per frame.
protocol AudioSource: AnyObject {
    /// Yields 320-sample Int16 frames as they arrive. Finishes when `stop()` is called.
    func frames() -> AsyncStream<[Int16]>
    func stop()
}

protocol AudioSink: AnyObject {
    /// Must be called before `write`. Idempotent.
    func start()
    func write(_ frame: [Int16]) async
    func stop()
}
